### PR TITLE
fix(core): don't leak sidecar data memory

### PR DESCRIPTION
**Pull request by mmastrac**
Split the sidecar data into a separate part of the output snapshot data so we don't need to put that data into the snapshot, which potentially leaks it. 

This also opens up the ability for us to have statically-referenced data in the sidecar snapshot which will make for some zero-copy features.

### CODE REVIEW

File: core/runtime/jsruntime.rs [hunk 1]
```diff
@@ -735,9 +731,10 @@ impl JsRuntime {
       );
 
       // Get module map data from the snapshot
-      if init_mode.is_from_snapshot() {
-        snapshotted_data =
-          Some(snapshot_util::get_snapshotted_data(scope, context));
+      if let Some(raw_data) = sidecar_data {
+        snapshotted_data = Some(snapshot::load_snapshotted_data_from_snapshot(
+          scope, context, raw_data,
+        ));
```
Comment (bartlomieju) [inline core/runtime/jsruntime.rs:737]: Yeah, this is cleaner

File: core/runtime/jsruntime.rs [hunk 2]
```diff
@@ -1856,7 +1853,8 @@ impl JsRuntimeForSnapshot {
     }
 
     // Serialize the module map and store its data in the snapshot.
-    {
+    // TODO(mmastrac): This should deconstruct the realm completely.
```
Comment (bartlomieju) [inline core/runtime/jsruntime.rs:1856]: I don't really understand this comment

File: core/runtime/snapshot.rs [hunk 3]
```diff
@@ -0,0 +1,577 @@
...
+  fn process_chunk(&mut self, chunk: &[u8]) -> std::io::Result<()> {
+    self.file.write_all(chunk)
+  }
+
+  fn finalize(self) -> std::io::Result<Self::Output> {
+    Ok(self.file)
+  }
+}
+
+/// Serializes the snapshot to another serializer, with a bulk compression operation performed on top.
```
Comment (bartlomieju) [inline core/runtime/snapshot.rs:245]: What is bulk compression in this context?

File: core/runtime/snapshot.rs [hunk 4]
```diff
@@ -0,0 +1,577 @@
...
+}
+
+pub(crate) struct SnapshottedData {
+  pub module_map_data: ModuleMapSnapshotData,
+  pub js_handled_promise_rejection_cb: Option<v8::Global<v8::Function>>,
+  pub ext_source_maps: HashMap<String, Vec<u8>>,
+}
+
+#[derive(Serialize, Deserialize)]
+pub(crate) struct RawSnapshottedData {
```
Comment (bartlomieju) [inline core/runtime/snapshot.rs:456]: Not a fan of this name - I don't see what's "raw" about this data - maybe `AuxSnapshotData`?

**Review by bartlomieju — APPROVED** [review]

File: core/runtime/snapshot.rs (hunk 3 above)
Comment (mmastrac) [reply core/runtime/snapshot.rs:260]: I stole the term from `zstd` which we use in the LSP snapshot: `&zstd::bulk::compress(snapshot_slice, 22)`

File: core/runtime/snapshot.rs (hunk 4 above)
Comment (mmastrac) [reply core/runtime/snapshot.rs:471]: I can rename this

File: core/runtime/jsruntime.rs (hunk 2 above)
Comment (mmastrac) [reply core/runtime/jsruntime.rs]: Expanded a bit

File: core/runtime/snapshot.rs (hunk 4 above)
Comment (mmastrac) [reply core/runtime/snapshot.rs]: This is now `SerializableSnapshotSidecarData` to indicate that its intended destination is serde.

**Pull request merged**